File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJVariableAssignmentsPsiUtil.java
package org.cappuccino_project.ide.intellij.plugin.psi.utils;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import org.cappuccino_project.ide.intellij.plugin.psi.*;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJCompositeElement;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJVariableAssignment;
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ObjJVariableAssignmentsPsiUtil {

    public static List<ObjJVariableAssignment> getAllVariableAssignmentsMatchingName(@NotNull final PsiElement element, @NotNull final String fqName) {
        List<ObjJVariableAssignment> bodyVariableAssignments = getAllVariableAssignements(element);
        return ArrayUtils.filter(bodyVariableAssignments, (assignment) -> {
            for (ObjJQualifiedReference qualifiedReference : assignment.getQualifiedReferenceList()) {
                if (qualifiedReference.getVariableNameList().isEmpty()) {
                    continue;
                }
                String currentVariableFqName = ObjJVariableNameUtil.getQualifiedNameAsString(qualifiedReference, null);
                if (Objects.equals(fqName, currentVariableFqName)) {
                    return true;
                }
            }
            return false;
        });
    }

    public static List<ObjJVariableAssignment> getAllVariableAssignements(PsiElement block) {
        List<ObjJCompositeElement> compositeElements = ObjJBlockPsiUtil.getParentBlockChildrenOfType(block, ObjJCompositeElement.class, true);
        List<ObjJVariableAssignment> declarations = new ArrayList<>();
        for (ObjJCompositeElement compositeElement : compositeElements) {
            if (compositeElement instanceof ObjJBodyVariableAssignment) {
                addVariableDeclarationFromBodyVariableAssignment(declarations,(ObjJBodyVariableAssignment)compositeElement);
            } else if (compositeElement instanceof ObjJExpr) {
                addVariableDeclarationFromExpression(declarations, (ObjJExpr)compositeElement);
            }
        }
        addVariableDeclarationsInFile(declarations, block.getContainingFile());
        return declarations;
    }

    private static void addVariableDeclarationsInFile(@NotNull List<ObjJVariableAssignment> declarations, @NotNull PsiFile file) {
        for (ObjJBodyVariableAssignment variableAssignment : ObjJTreeUtil.getChildrenOfTypeAsList(file, ObjJBodyVariableAssignment.class)) {
            addVariableDeclarationFromBodyVariableAssignment(declarations, variableAssignment);
        }
    }

    private static void addVariableDeclarationFromBodyVariableAssignment(@NotNull List<ObjJVariableAssignment> declarations, @NotNull ObjJBodyVariableAssignment bodyVariableAssignment) {
        declarations.addAll(ObjJTreeUtil.getChildrenOfTypeAsList(bodyVariableAssignment, ObjJVariableAssignment.class));
    }

    private static void addVariableDeclarationFromExpression(@NotNull List<ObjJVariableAssignment> declarations, @NotNull ObjJExpr expression) {
        if (expression.getLeftExpr() == null) {
            return;
        }
        if (expression.getLeftExpr().getVariableDeclaration() == null) {
            return;
        }
        declarations.addAll(ObjJTreeUtil.getChildrenOfTypeAsList(expression.getLeftExpr(), ObjJVariableAssignment.class));
    }

    @NotNull
    public static ObjJExpr getAssignedValue(@NotNull ObjJVariableAssignmentLogical assignmentLogical) {
        return assignmentLogical.getAssignmentExprPrime().getExpr();
    }

}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJVariableAssignmentsPsiUtil.kt
package org.cappuccino_project.ide.intellij.plugin.psi.utils

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import org.cappuccino_project.ide.intellij.plugin.psi.*
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJCompositeElement
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJVariableAssignment
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils

import java.util.ArrayList

object ObjJVariableAssignmentsPsiUtil {

    fun getAllVariableAssignmentsMatchingName(element: PsiElement, fqName: String): List<ObjJVariableAssignment> {
        val bodyVariableAssignments = getAllVariableAssignements(element)
        return ArrayUtils.filter(bodyVariableAssignments) filter@ { assignment ->
            for (qualifiedReference in assignment.qualifiedReferenceList) {
                if (qualifiedReference.variableNameList.isEmpty()) {
                    continue
                }
                val currentVariableFqName = ObjJVariableNameUtil.getQualifiedNameAsString(qualifiedReference, null)
                if (fqName == currentVariableFqName) {
                    return@filter true
                }
            }
            false
        }
    }

    fun getAllVariableAssignements(block: PsiElement): List<ObjJVariableAssignment> {
        val compositeElements = block.getParentBlockChildrenOfType(ObjJCompositeElement::class.java, true)
        val declarations = ArrayList<ObjJVariableAssignment>()
        for (compositeElement in compositeElements) {
            if (compositeElement is ObjJBodyVariableAssignment) {
                addVariableDeclarationFromBodyVariableAssignment(declarations, compositeElement)
            } else if (compositeElement is ObjJExpr) {
                addVariableDeclarationFromExpression(declarations, compositeElement)
            }
        }
        addVariableDeclarationsInFile(declarations, block.containingFile)
        return declarations
    }

    private fun addVariableDeclarationsInFile(declarations: MutableList<ObjJVariableAssignment>, file: PsiFile) {
        for (variableAssignment in file.getChildrenOfType( ObjJBodyVariableAssignment::class.java)) {
            addVariableDeclarationFromBodyVariableAssignment(declarations, variableAssignment)
        }
    }

    private fun addVariableDeclarationFromBodyVariableAssignment(declarations: MutableList<ObjJVariableAssignment>, bodyVariableAssignment: ObjJBodyVariableAssignment) {
        declarations.addAll(bodyVariableAssignment.getChildrenOfType( ObjJVariableAssignment::class.java))
    }

    private fun addVariableDeclarationFromExpression(declarations: MutableList<ObjJVariableAssignment>, expression: ObjJExpr) {
        if (expression.leftExpr == null) {
            return
        }
        if (expression.leftExpr == null) {
            return
        }
        declarations.addAll(expression.leftExpr!!.variableDeclaration!!.getChildrenOfType(ObjJVariableAssignment::class.java))
    }

    fun getAssignedValue(assignmentLogical: ObjJVariableAssignmentLogical): ObjJExpr {
        return assignmentLogical.assignmentExprPrime.expr
    }

}
